### PR TITLE
Replaced explicit __metaclass__ assignment with @six.add_metaclass

### PR DESCRIPTION
**Pull request by superbobry**
`__metaclass__ `only works for Python 2. In Python 3 it has no effect since the metaclass syntax is

    class A(meta=abc.ABCMeta):
        ...

`@six.add_metaclass` covers both variants.

### CODE REVIEW

File: tensorflow/contrib/learn/python/learn/estimators/estimator.py [hunk 1]
```diff
@@ -403,8 +403,6 @@ class BaseEstimator(sklearn.BaseEstimator, evaluable.Evaluable,
   Users should not instantiate or subclass this class. Instead, use an
   `Estimator`.
   """
-  __metaclass__ = abc.ABCMeta
```
Comment (superbobry) [inline tensorflow/contrib/learn/python/learn/estimators/estimator.py]: I did not decorate this class with `@six.add_metaclass`, because its base classes already have `abc.ABCMeta` as the metaclass.

**Comment by drpngx** [comment]
I'm not familiar enough with that feature. @allenlavoie do you?

It begs the question, since we had this for a long time, why we don't test for it, and why no one noticed. If it isn't useful I would vote for removing it altogether, or if it does, then testing for it.

**Review by allenlavoie — DISMISSED** [review]
Thank you for doing this!

As for why nobody noticed that ABCMeta wasn't added in Python 3: its effects are relatively subtle (exceptions on instantiation rather than on abstract method calls, and enabling abstract properties which are rare). We also mostly develop in Python 2 (unfortunately), and ABCMeta only comes up at runtime if something is wrong with inheritance (which we hopefully work out before committing).

So I agree that having a largely useless metaclass registered is a bit odd, but it's pretty idiomatic too. I'd look at it as similar to type annotations.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @drpngx: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by drpngx** [comment]
@superbobry sorry this dropped out of the radar. Could you pull rebase and push again?

**Comment by superbobry** [comment]
@drpngx done

**Comment by superbobry** [comment]
I think the ``tensorflow/tools/api/tests:api_compatibility_test`` failure is harmless. The classes do indeed lack the ``__metaclass`` attribute, because the metaclass [is applied](https://github.com/benjaminp/six/blob/68112f3193c7d4bef5ad86ed1b6ed528edd9093d/six.py#L835-L848) via a decorator, but this is not a breaking change.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @drpngx: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @drpngx: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by drpngx** [comment]
Sorry this dropped off the radar. Testing...

**Comment by drpngx** [comment]
Failures seem unrelated, testing again.

**Comment by yifeif** [comment]
Seems to break api_compatibility test. Should the golden files be updated?

**Comment by tensorflowbutler** [comment]
Nagging Assignee @drpngx: It has been 20 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @drpngx: It has been 35 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @drpngx: It has been 50 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by drpngx** [comment]
@superbobry could you update the goldens?

**Comment by martinwicke** [comment]
This is a worthwhile change, but yes, please regenerate the golden files to reflect the changes.

**Comment by martinwicke** [comment]
I triggered the tests which should fail and leave a log containing how to update the goldens.

**Comment by drpngx** [comment]
```
ERROR:tensorflow:19 differences found between API and golden.
Issue 1	: Object tensorflow.linalg.LinearOperatorCirculant2D.__metaclass__ expected but not found (removed).
Issue 2	: Object tensorflow.linalg.LinearOperatorDiag.__metaclass__ expected but not found (removed).
Issue 3	: Object tensorflow.linalg.LinearOperatorIdentity.__metaclass__ expected but not found (removed).
Issue 4	: Object tensorflow.linalg.LinearOperator.__metaclass__ expected but not found (removed).
Issue 5	: Object tensorflow.linalg.LinearOperatorScaledIdentity.__metaclass__ expected but not found (removed).
Issue 6	: Object tensorflow.linalg.LinearOperatorCirculant.__metaclass__ expected but not found (removed).
Issue 7	: Object tensorflow.linalg.LinearOperatorCirculant3D.__metaclass__ expected but not found (removed).
Issue 8	: 'path: "tensorflow.data.Dataset"\ntf_class {\n  is_instance: "<class \\\'tensorf [truncated]... != 'path: "tensorflow.data.Dataset"\ntf_class {\n  is_instance: "<class \\\'tensorf [truncated]...
  path: "tensorflow.data.Dataset"
  tf_class {
    is_instance: "<class \'tensorflow.python.data.ops.dataset_ops.Dataset\'>"
    is_instance: "<type \'object\'>"
    member {
      name: "output_classes"
      mtype: "<class \'abc.abstractproperty\'>"
    }
    member {
      name: "output_shapes"
      mtype: "<class \'abc.abstractproperty\'>"
    }
    member {
      name: "output_types"
      mtype: "<class \'abc.abstractproperty\'>"
    }
    member_method {
      name: "__init__"
-     argspec: "args=[\'self\'], varargs=None, keywords=None, defaults=None"
    }
    member_method {
      name: "apply"
      argspec: "args=[\'self\', \'transformation_func\'], varargs=None, keywords=None, defaults=None"
    }
    member_method {
      name: "batch"
      argspec: "args=[\'self\', \'batch_size\', \'drop_remainder\'], varargs=None, keywords=None, defaults=[\'False\'], "
    }
    member_method {
      name: "cache"
      argspec: "args=[\'self\', \'filename\'], varargs=None, keywords=None, defaults=[\'\'], "
    }
    member_method {
      name: "concatenate"
      argspec: "args=[\'self\', \'dataset\'], varargs=None, keywords=None, defaults=None"
    }
    member_method {
      name: "filter"
      argspec: "args=[\'self\', \'predicate\'], varargs=None, keywords=None, defaults=None"
    }
    member_method {
      name: "flat_map"
      argspec: "args=[\'self\', \'map_func\'], varargs=None, keywords=None, defaults=None"
    }
    member_method {
      name: "from_generator"
      argspec: "args=[\'generator\', \'output_types\', \'output_shapes\', \'args\'], varargs=None, keywords=None, defaults=[\'None\', \'None\'], "
    }
    member_method {
      name: "from_sparse_tensor_slices"
      argspec: "args=[\'sparse_tensor\'], varargs=None, keywords=None, defaults=None"
    }
    member_method {
      name: "from_tensor_slices"
      argspec: "args=[\'tensors\'], varargs=None, keywords=None, defaults=None"
    }
    member_method {
      name: "from_tensors"
      argspec: "args=[\'tensors\'], varargs=None, keywords=None, defaults=None"
    }
    member_method {
      name: "interleave"
      argspec: "args=[\'self\', \'map_func\', \'cycle_length\', \'block_length\', \'num_parallel_calls\'], varargs=None, keywords=None, defaults=[\'1\', \'None\'], "
    }
    member_method {
      name: "list_files"
      argspec: "args=[\'file_pattern\', \'shuffle\', \'seed\'], varargs=None, keywords=None, defaults=[\'None\', \'None\'], "
    }
    member_method {
      name: "make_initializable_iterator"
      argspec: "args=[\'self\', \'shared_name\'], varargs=None, keywords=None, defaults=[\'None\'], "
    }
    member_method {
      name: "make_one_shot_iterator"
      argspec: "args=[\'self\'], varargs=None, keywords=None, defaults=None"
    }
    member_method {
      name: "map"
      argspec: "args=[\'self\', \'map_func\', \'num_parallel_calls\'], varargs=None, keywords=None, defaults=[\'None\'], "
    }
    member_method {
      name: "padded_batch"
      argspec: "args=[\'self\', \'batch_size\', \'padded_shapes\', \'padding_values\', \'drop_remainder\'], varargs=None, keywords=None, defaults=[\'None\', \'False\'], "
    }
    member_method {
      name: "prefetch"
      argspec: "args=[\'self\', \'buffer_size\'], varargs=None, keywords=None, defaults=None"
    }
    member_method {
      name: "range"
      argspec: "args=[], varargs=args, keywords=None, defaults=None"
    }
    member_method {
      name: "reduce"
      argspec: "args=[\'self\', \'initial_state\', \'reduce_func\'], varargs=None, keywords=None, defaults=None"
    }
    member_method {
      name: "repeat"
      argspec: "args=[\'self\', \'count\'], varargs=None, keywords=None, defaults=[\'None\'], "
    }
    member_method {
      name: "shard"
      argspec: "args=[\'self\', \'num_shards\', \'index\'], varargs=None, keywords=None, defaults=None"
    }
    member_method {
      name: "shuffle"
      argspec: "args=[\'self\', \'buffer_size\', \'seed\', \'reshuffle_each_iteration\'], varargs=None, keywords=None, defaults=[\'None\', \'None\'], "
    }
    member_method {
      name: "skip"
      argspec: "args=[\'self\', \'count\'], varargs=None, keywords=None, defaults=None"
    }
    member_method {
      name: "take"
      argspec: "args=[\'self\', \'count\'], varargs=None, keywords=None, defaults=None"
    }
    member_method {
      name: "window"
      argspec: "args=[\'self\', \'size\', \'shift\', \'stride\', \'drop_remainder\'], varargs=None, keywords=None, defaults=[\'None\', \'1\', \'False\'], "
    }
    member_method {
      name: "zip"
      argspec: "args=[\'datasets\'], varargs=None, keywords=None, defaults=None"
    }
  }
Issue 9	: Object tensorflow.linalg.LinearOperatorZeros.__metaclass__ expected but not found (removed).
Issue 10	: Object tensorflow.data.Dataset.__metaclass__ expected but not found (removed).
Issue 11	: Object tensorflow.linalg.LinearOperatorLowRankUpdate.__metaclass__ expected but not found (removed).
Issue 12	: Object tensorflow.data.FixedLengthRecordDataset.__metaclass__ expected but not found (removed).
Issue 13	: Object tensorflow.linalg.LinearOperatorBlockDiag.__metaclass__ expected but not found (removed).
Issue 14	: Object tensorflow.linalg.LinearOperatorKronecker.__metaclass__ expected but not found (removed).
Issue 15	: Object tensorflow.linalg.LinearOperatorFullMatrix.__metaclass__ expected but not found (removed).
Issue 16	: Object tensorflow.data.TFRecordDataset.__metaclass__ expected but not found (removed).
Issue 17	: Object tensorflow.linalg.LinearOperatorComposition.__metaclass__ expected but not found (removed).
Issue 18	: Object tensorflow.data.TextLineDataset.__metaclass__ expected but not found (removed).
Issue 19	: Object tensorflow.linalg.LinearOperatorLowerTriangular.__metaclass__ expected but not found (removed).
```

**Comment by drpngx** [comment]
@superbobry could you update the goldens?

**Review by martinwicke — APPROVED** [review]
All introduced API change appear to be nominal.

**Comment by superbobry** [comment]
Hm, `api_compatibility_test` passes locally, and I don't seem to have `tf.data.experimental`:

```bash
$ ls tensorflow/tools/api/golden/v*/tensorflow.data.*                                                                                                          
tensorflow/tools/api/golden/v1/tensorflow.data.-dataset.pbtxt                     tensorflow/tools/api/golden/v2/tensorflow.data.-dataset.pbtxt
tensorflow/tools/api/golden/v1/tensorflow.data.-fixed-length-record-dataset.pbtxt tensorflow/tools/api/golden/v2/tensorflow.data.-fixed-length-record-dataset.pbtxt
tensorflow/tools/api/golden/v1/tensorflow.data.-iterator.pbtxt                    tensorflow/tools/api/golden/v2/tensorflow.data.-iterator.pbtxt
tensorflow/tools/api/golden/v1/tensorflow.data.-t-f-record-dataset.pbtxt          tensorflow/tools/api/golden/v2/tensorflow.data.-t-f-record-dataset.pbtxt
tensorflow/tools/api/golden/v1/tensorflow.data.-text-line-dataset.pbtxt           tensorflow/tools/api/golden/v2/tensorflow.data.-text-line-dataset.pbtxt
tensorflow/tools/api/golden/v1/tensorflow.data.pbtxt
```

**Comment by josh11b** [comment]
Fine for API review.

**Comment by superbobry** [comment]
Yes, I ran on Python 2.7.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @drpngx: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by superbobry** [comment]
To be submitted internally.
